server/graphql/v2/object/Expense.js

```diff
@@ -42,6 +45,18 @@ const Expense = new GraphQLObjectType({
           };
         },
       },
+      payee: {
+        type: Account,
+        description: 'The account being paid by this expense',
+        async resolve(expense, _, req) {
+          // Set the permissions for account's fields
+          const canSeePrivateInfo = (await canViewExpensePrivateInfo(expense, req)).userLocation;
+          setContextPermission(req, PERMISSION_TYPE.SEE_ACCOUNT_LOCATION, expense.FromCollectiveId, canSeePrivateInfo);
```

Comment:
Not sure how useful is this, couldn't we just use `canViewExpensePrivateInfo` directly?

Reply:
Some permissions depend on the context. For example, you're normally not allowed to watch the address of a user. But if you're a host admin validating an expense that the user has submitted, then you're allowed to see it.

Problem is: we have no clue in the User's location resolver about this context. We could eventually go up in the AST from there to get it, but it would get pretty messy quickly (and I'd argue that the user location resolver should not have any logic related to expenses).

So how to represent this kind of permissions check that depends on the parent?

For the incognito feature the decision was to store it in the collective as a virtual field (`inTheContextOfUserId`). We didn't block this decision because the feature needed to be released fast, yet it is problematic: there's no reason for the collective model to know about Graphql's context and anyone anywhere can change this value with whatever they want.

With this small helper, what I tried to do is to store this information in a standardize way in the GraphQl context. I believe storing it there is right, because it's what this object is about: you normally can't access this data, but in the context of **this query** you're allowed to.

Do you see better ways to achieve that?

Reply:
Cool, I get it now why this is happening.
One alternative is to set the user role in the context and write the permission logic in the resolver.

What I mean here is that the context would hold and enum value related to its user role _e.g._ `ROLE.HOST_ADMIN` or `ROLE.PARENT_COLLECTIVE_ADMIN` or `ROLE.OWNER`.
In this scenario, the resolver would take that generic context information and be explicit about its access logic:
```ts
resolve(individual, _, req) {
  if (
    req.role === ROLE.HOST_ADMIN ||
    req.role === ROLE.OWNER
  ) {
    return individual.location;
  }
}
```
The idea would be to work on a more abstracted level than going straight to granular permissions.
WDYT?

Reply:
I could work with that approach, but the permissions here are too vague: 
- Host admin of which host? Owner of what? 
- It doesn't account for users with multiple roles (ie. owner of the expense **and** host admin)

Let's take a more complex query:
 
```
Host
└── Collective
  └── MemberCollectives
      └── Expenses
          ├── PayoutMethod => can only see if owner of PayoutMethod or host admin
          │   └── data
          └── User
              └── email => can only see if collective admin, owner or host admin
```

With your proposal, for example, how would you make sure that the `ROLE.HOST_ADMIN` was set for the proper collective? Or said another way, how would you know if the role was set at the `Collective` or at the `MemberCollective` level?

We could address this question by adding something like:

```ts
// Data
req.roles = {
  Expense: {
    42: ['ADMIN', 'OWNER']
  },
  Collective: {
    56: ['ADMIN']
  }
}

// Lib
type Entity = 'Expense' | 'Colllective' | ...
type Role = 'ADMIN' | 'OWNER'
setContextRole(req, entity: Entity, entityId, role: Role) { ... }
hasContextRole(req, entity, entityId) {...}
```

But then we go back to the initial problem: the user's resolver at the bottom of the tree doesn't know which expense it was retrieved from.

Reply:
I could work with that approach, but the permissions here are too vague: 
- Host admin of which host? Owner of what? 
- It doesn't account for users with multiple roles (ie. owner of the expense **and** host admin)

Let's take a more complex query:
 
```
Host
└── Collective
  └── MemberCollectives
      └── Expenses
          ├── PayoutMethod => can only see if owner of PayoutMethod or host admin
          │   └── data
          └── User
              └── email => can only see if collective admin, owner or host admin
```

With your proposal, for example, how would you make sure that the `ROLE.HOST_ADMIN` was set for the proper collective? Or said another way, how would you know if the role was set at the `Collective` or at the `MemberCollective` level?

We could address this question by adding something like:

```ts
// Data
req.roles = {
  Expense: {
    42: ['ADMIN', 'OWNER']
  },
  Collective: {
    56: ['ADMIN']
  }
}

// Lib
type Entity = 'Expense' | 'Colllective' | ...
type Role = 'ADMIN' | 'OWNER'
setContextRole(req, entity: Entity, entityId, role: Role) { ... }
hasContextRole(req, entity, entityId) {...}
```

But then we go back to the initial problem: the user's resolver at the bottom of the tree doesn't know which expense it was retrieved from.

Reply:
Another example to demonstrate that it is actually dangerous:

```
Collective
└── RelatedCollective
    └── private
```

If `RelatedCollective` resolver forgets to set `req.roles = COLLECTIVE_ADMIN`, then you'll get access to `private` without being an admin of RelatedCollective.

Also, this needs to be verified, but if resolvers can run in parallel (which I assume they do when they're asynchronous) then we can't trust that `req.roles` because we won't be able to know which part of the code filled it.

Reply:
I'm merging to we can move forward, feel free to open an issue on this topic if you want to explore further alternatives

Reply:
Sure! This is a wicked problem that requires hard thinking.
I'm fine with the current implementation, perhaps it is just a bit complex because the problem is also complex.